Chain Name Service Draft (#100)
tendermint/spn: docs/cns.md

# Chain Name Service Draft

The goal of this module is to be a simple registry with the ability to transfer ownership. Auctions and other mechanisms are out of scope and can be implemented alongside or on top of this system (perhaps, in a custodial way).

# State

> ## Chain name

Reviewer: ```suggestion
## Chain Info
```

> Chain name contains information about the chain, most importantly, chain ID used to uniquly identify the chain and a list of peer nodes.

Reviewer 2: Chain Name is too confusing with Chain-ID imo, I think it can be renamed to `ChainInfo`?

> Chain name contains information about the chain, most importantly, chain ID used to uniquly identify the chain and a list of peer nodes.

Reviewer 2: ```suggestion
Chain name contains information about the chain, most importantly, chain ID used to uniquely identify the chain and a list of peer nodes.
```

> Chain name contains information about the chain, most importantly, chain ID used to uniquly identify the chain and a list of peer nodes.

Reviewer: ```suggestion
Chain Info contains information about the chain, most importantly, chain ID used to uniquly identify the chain and a list of peer nodes.
```

> Key: `chain-[chainID] -> Chain Name` 

Reviewer 2: Why does this key need to be prefixed if this is the only chains are being stored by the registry?

Reviewer 3: This is the prefix from the root of the module key. I think it's good to always have a prefix for a specific entry in the store even if it's the only data for the module.

> Key: `chain-[chainID] -> Chain Name` 

Reviewer: ```suggestion
Key: `[chainID] -> Chain Info` 
```

```go

> type ChainName struct {

Reviewer: ```suggestion
type ChainInfo struct {
```

	ChainID string

> 	Owner sdk.AccAddress

Reviewer 2: Given these might eventually be controlled over IBC, this should be a string. The counterparty chains may not have a notion of sdk.AccAddress. The owner may be a channel/port pair.

Reviewer: If it were controlled over IBC it would still use an Interchain Account which is represented by `sdk.AccAddress`, right? Is there another format where an IBC channel/port pair would be preferable?

> 	Peers []peer

Reviewer: are these all peers or just seed nodes?

Reviewer 4: Add `[]seed`

Author: Yes, I think seed nodes.

> 	Peers []peer

Reviewer: ```suggestion
	Seeds []seed
```

> 	SourceURL string

Reviewer: what is this for?

Reviewer 4: Ideally this is an `[]net.URL` of open RPC endpoints.

Author: It's the URL of the source code from which the chain is built.

Reviewer 5: I think it would also be beneficial for light clients and state sync that trusted RPC endpoints are added here

Reviewer: ```suggestion
	SourceCodeURL string
```

Reviewer: @cmwaters brings up an important question about the function of the registry:

Should it be a place with infrequent manual updates (seednodes, sourceCodeURL, genesis file) or with frequent manual updates (trusted RPC endpoints). I think it would be good getting some estimates about the expected costs for maintaining a list of RPC endpoints. @alexanderbez is almost done with the node crawler feature of https://github.com/cosmos/atlas. I'm sure we'll be able to get a lot of info of the rate that changes take place from here and can multiply by tx gas estimates to get ar ough idea of how feasible it is to update RPC endpoints. I'd argue this opens up a much higher risk of security exploits occurring as well, as verifying RPC endpoints to be valid is difficult if not impossible as far as I understand.

Reviewer 5: True. Trusted is probably not the right word and this is also more a thing of convenience, a nice to have, then something that is core to the chain registry. I have just heard in a few separate instances that finding a good set of RPC endpoints is a bit of a pain point for new users. This might be redundant if crawling is effective enough or if light client and state sync can also work via p2p

> 	SourceCommitHash string

Reviewer: what is this?

Reviewer 4: I think hes saying genesis file location?

Author: Commit hash of the source code.

Reviewer 5: This could also be the hash of the binary (but this is much the same as commit hash)

Reviewer 5: I would also think it is necessary to add a genesis hash

Reviewer: I could imagine a genesis hash is more valuable than a binary hash. There would ideally be multiple binaries for operating the node anyway, or some customization at a non-state-breaking level. The genesis hash needs to be consistent across them though.

Reviewer 6: Agree having genesis hash is more useful, but I also feel it's better to include a commit hash as going forward chains using upgrade module will not have genesis files being changed but just the binaries

> 	SourceCommitHash string

Reviewer 2: ```suggestion
	SourceCommitHash string
	CanonicalIBCClientId string
```

> 	SourceCommitHash string

Reviewer: ```suggestion
	GenesisHash string
```

}
```

## Offer

Offers are used to transfer ownership of a chain name.

```go
type Offer struct {

> 	ID int32

Reviewer: ```suggestion
	ID int32
	ChainId string
```

	From sdk.AccAddress
	Amount sdk.Coins

> 	Status status // pending | rejected | approved

Reviewer: offer doesn't point to any specific chain-id. There doesn't seem to be any sort of join data structure which would link them either.

}
```

# Messages

## MsgEditChainName

Chain owner can make edits to the chain name.

```go

> type MsgEditChainName struct {

Reviewer 2: It would be nice to have all of these msgs implemented as IBCPackets so these can be sent from the domain-controlling chain to the registry chain

Reviewer: isn't this possible already with interchain accounts?

	
}
```

## MsgTransferChainName

> To transfer a name from one owner from another without creating an offer with a zero amount.

Reviewer: ```suggestion
To transfer a name from one owner to another without creating an offer with a zero amount.
```

Reviewer: why would this require not making an offer with a zero amount? does it require making an offer with non-zero amount or not making an offer at all?

```go
type MsgTransferChainName struct {

> 	From string

Reviewer: ```suggestion
	From sdk.AccAddress
```
right?

> 	To string

Reviewer: ```suggestion
	To sdk.AccAddress
```

	ChainID string
}
```

## MsgCreateOffer

```go
type MsgCreateOffer struct {
	ChainID string
	Amount sdk.Coins
}
```

## MsgRejectOffer

```go
type MsgRejectOffer struct {

> 	ChainID string

Reviewer: ```suggestion
	ID int32
	ChainID string
```

}
```

## MsgApproveOffer

```go
type MsgApproveOffer struct {

> 	ChainID string

Reviewer: ```suggestion
	ID int32
	ChainID string
```
same as last one, actually chainID becomes unnecessary but probably good to ensure there's not incorrect target of offer ID

}
```

## MsgRemoveOffer

```go
type MsgRejectOffer struct {

> 	ChainID string

Reviewer: ```suggestion
	ID int32
	ChainID string
```

}
```

## MsgRegisterName

> To accuire a name that doesn't have an owner. By default each name has a price. When registering a name, amount equivalent to the price goes to the community pool.

Reviewer: ENS uses an oracle now to ensure price stability. Until that feature was available they just locked ETH and gave it back once the name was revoked. Maybe that's a way that would be simple for this to begin? It would be unfortunate if early users were punished by paying in ATOM instead of the fiat equivalent at the time. On the other hand going to the community pool is pretty easy to justify.

Reviewer 4: What data does the oracle bring in? Also I'm in favor of community pool.

Reviewer: I think an oracle might be valuable in other areas more than here (confirming seednodes, potentially RPC endpoints) but in this case it was a USD/Crypto oracle so that prices of new names weren't volatile with the underlying crypto used to claim them.

> ```go
> type MsgRejectOffer struct {
> 	ChainID string
> }
> ```

Reviewer: ```suggestion
```
redundant

# Ideas

- Reserve names for existing chains

> - Names with a chain IDs that have `tesnet-` prefix are registered for free

Reviewer: ```suggestion
- Names with chain IDs that have `testnet-` prefix are registered for free
```

> - Off-chain tool to check that nodes in peer list are part of a network with a given chain ID. Can loop through all the chain names and create a table.

Reviewer: This is an oracle's job for constant updates. Maybe it can be combined with the Gravity Bridge (peggy) since the validator set will essentially be a functional oracle by operating the bridge.

As a short term solution I'd suggest maintaining a list of seed nodes would be sufficient since they can be used to expose all other nodes. Otherwise there's a potentially huge cost in keeping constantly updated a list of active nodes beyond the seed list.

Reviewer 4: We also have a generalized oracle module I'm working on...

Reviewer 3: > We also have a generalized oracle module I'm working on...

It would be a great solution.
I think we should distinguish in the chainInfo, the info that can be arbitrarily decided by the chain owning entity (source code URL, canonical IBC client ID) and what can be trustlessly verified (seed list, genesis hash?) and having this managed by an oracle system

> - If you're launching a chain and don't want a single entity to be responsible for managing a chain name, use a multisig

Reviewer: could be good enough. Security is really only part of the consideration. IBC is providing the app hash of all potential chain-ids already. I think the goal should be to leverage the actual connections as the source of updated information

> - This should probably be an IBC module

Reviewer: the IBC keeper could be imported so that state could be read. Maybe eventually an endblocker to check if any IBC connection got updated with that chain id?

Reviewer 4: I think this would be a cool first usecase for interchain accounts potentially?

> - Open question: how does relate to IBC, relayers. How can this module be used with IBC.

Reviewer: ah, began addressing above. Will copy past previous comment here:
> IBC is providing the app hash of all potential chain-ids already. I think the goal should be to leverage the actual connections as the source of updated information.

I think first step should be to replicate what the github is trying to do right now:
 * Chain-ID
 * Seed Nodes
 * Whitelist for administration of those values

I think second step should be some way to associate a chain ID with a list of connection IDs (since there can be multiple connections between the same chain). This is a tricky oracle problem. If we're trusting the "admins" of the chain id, we could extend that to their judgement of which connections are legitimate. There should be a way to confirm a second connection once a first one is made as well but I'm not sure exactly what steps are needed.

For simplicity maybe admin can approve manually is a good first step though, and it's just weak subjectivity as a guarantee?

ApeUnit is working on the flow for a wallet or block explorer to confirm an asset displayed with an account query by utilizing the github + tmcrawl/atlas at the moment. I think this should be the initial goal for the chain-registry as well.
